[WTF] Implement "reduce" function for Vector

mdubet opened the pull request:
#### 934ab5ee72c37c5c40e9b07274f538c5ec89c01a
<pre>
[WTF] Implement &quot;reduce&quot; function for Vector
<a href="https://bugs.webkit.org/show_bug.cgi?id=246140">https://bugs.webkit.org/show_bug.cgi?id=246140</a>

Reviewed by NOBODY (OOPS!).

This is a very basic implementation of a reduce/accumulate function
in a fold-left fashion.

* Source/WTF/wtf/Vector.h:
(WTF::Vector::reduce):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator&lt;&lt;):
(WebCore::reduce): Deleted.
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
(TestWebKitAPI::TEST):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/934ab5ee72c37c5c40e9b07274f538c5ec89c01a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91964 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22501 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101785 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161854 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/95966 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1204 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29650 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84436 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97991 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97622 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/739 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78514 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27691 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82661 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70727 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83324 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36055 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16281 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78380 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33793 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17383 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27051 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37670 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40097 "Found 2 new test failures: imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.html?h264_annexb, imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.worker.html?h264_annexb (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81002 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39554 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36507 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17797 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode (failure)") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/Vector.h:
```diff
@@ -903,6 +904,20 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
     template<typename ResultVector, typename MapFunction>
     auto map(MapFunction&&) const -> std::enable_if_t<std::is_invocable_v<MapFunction, const T&>, ResultVector>;
 
+    using BinOp = std::function<T(T, T)>;
```
It may be better for this function to be templated and for the BinOp to be a template parameter, as is usually done on Vector functions. This would avoid a potentially heap allocation inside the std::function constructor.

Also note that it seems the current std::function could cause unnecessary copying of the 2 function parameters (`T` vs `const T&`), which would matter for types like String.

cdumez replied:
We try to avoid abbreviations in WebKit, we should rename 'binOp / BinOp' to something better.

cdumez commented on Source/WTF/wtf/Vector.h:
```diff
@@ -903,6 +904,20 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
     auto map(MapFunction&&) const -> std::enable_if_t<std::is_invocable_v<MapFunction, const T&>, ResultVector>;
 
+    using BinOp = std::function<T(T, T)>;
+    T reduce(const BinOp& binOp, std::optional<T> initial = { })
```
Why is this function not const?

cdumez commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -60,48 +58,32 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
             return a + " " + b;
         };
-        ts << reduce(result, addWithSpace);
+        ts << result.reduce(addWithSpace);
```
Sadly, I don't think this is a great use case for your reduce implementation. Your implementation is unnecessarily inefficient for Strings and a StringBuilder is recommended for such operations.

cdumez replied:
If you really want to add reduce() to Vector, I recommend finding a better call site for it.

For this call site, we should be using something like this (which we could put in WTFString.h to promote reusability):
```
static String makeStringByJoining(const Vector<String>& strings, UChar separator)
{
    StringBuilder builder;
    for(auto& string : strings) {
        if (!builder.isEmpty())
            builder.append(separator);
        builder.append(string);
    }
    return builder.toString();
}
```

I believe this would be more efficient than your generic reduce implementation.

Or, even more efficient, we don't allocate a Vector at all:
```
StringBuilder builder;
auto appendToBuilder = [&](const String& string) {
    if (!builder.isEmpty())
        builder.append(' ');
    builder.append(string);
};
if (values.stylistic)
     appendToBuilder("stylistic(" + *values.stylistic + ")");
if (values.historicalForms)
    appendToBuilder("historical-forms"_s);
// ...
ts << builder.toString();
```

darinadler replied:
I strongly agree that we must not add code sites that use reduce in a way that results in making tons of temporary strings. And using a vector has the same problem.

For example:
```
appendToBuilder("stylistic(" + *values.stylistic + ")");
```
This is still not a good pattern. It creates and destroys a temporary string; for best efficiency we need each of the pieces to be appended to the builder individually without creating and destroying any temporaries.

For adding separators while using `StringBuilder`, I have been thinking about how to create an object that inserts no separator the first time, but then inserts the separator later. If we did create it, the idiom would be something like this.
```
SeparatorCharacter separator { ' ' };
if (values.stylistic)
    builder.append(separator, "stylistic(", *values.stylistic, ')');
if (values.historicalForms)
    builder.append(separator, "historical-forms");
// ...
```

The other thing I noticed at this call site is that we are mixing `TextStream` and `StringBuilder`. They are two competing interfaces that try to do the same kind of thing, a framework for serializing to strings. Not new to this patch, but vexing.

darinadler replied:
This idiom reads a little bit crazy, but would work today and be quite efficient:
```
const char* separator = "";
if (values.stylistic)
    builder.append(std::exchange(separator, " "), "stylistic(", *values.stylistic, ')');
if (values.historicalForms)
    builder.append(std::exchange(separator, " "), "historical-forms");
```

mdubet replied:
Yes indeed for this use case `reduce` is less than ideal performance-wise.
I've changed it to use lambda + variadic args, which I think avoid unnecessary string or vector creation/destruction.

darinadler reviewed (changes requested):
Great idea to have a reduce, but not an efficient idiom for string building.

mdubet committed:
[WTF] Implement "reduce" function for Vector
https://bugs.webkit.org/show_bug.cgi?id=246140

Reviewed by NOBODY (OOPS!).

This is a very basic implementation of a reduce/accumulate function
in a fold-left fashion.

* Source/WTF/wtf/Vector.h:
(WTF::Vector::reduce):
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator<<):
(WebCore::reduce): Deleted.
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
(TestWebKitAPI::TEST):

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -60,48 +58,35 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
+        StringBuilder builder;
+        auto add = [&builder](auto&& ...args) { 
+            if (!builder.isEmpty())
+                builder.append(" ");
```
Slightly more efficient to add ‘ ‘ rather than “ “.

darinadler commented on Source/WebCore/platform/text/TextFlags.cpp:
```diff
@@ -60,48 +58,35 @@ WTF::TextStream& operator<<(TextStream& ts, Kerning kerning)
+            if (!builder.isEmpty())
+                builder.append(" ");
+
+            builder.append(args...);
```
I suspect we need  to use `std::forward` here to avoid copying arguments? I’d like to confirm we are not creating and destroying copies of `String` objects.

darinadler commented on Source/WTF/wtf/Vector.h:
```diff
@@ -903,6 +904,19 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
     template<typename ResultVector, typename MapFunction>
     auto map(MapFunction&&) const -> std::enable_if_t<std::is_invocable_v<MapFunction, const T&>, ResultVector>;
 
+    template <typename BinaryOperation> T reduce(const BinaryOperation& binaryOperation, std::optional<T> initial = { }) const
```
Please match the formatting used just above without the space after `template`. Ideally we would be consistent across all of WebKit but at least let’s stay locally consistent.

Can we find at least one place to use this besides the test? Normally we would not add something unused.

I’d prefer to have a more general purpose algorithm so it’s just as easy to use this on a span as on a vector, or on a `ListHasSet` or even a `HashSet`. So maybe it should be a non-member function. We don’t want to create more functions that lead people to unnecessarily create temporary vectors when what they want to operate on is already in some other kind of array or a subset of a vector or any other collection.

cdumez replied:
Why add reduce() at all if there is no call site for it? I'd rather we introduce the reduce() function with some code calling it to make sure it is actually used and useful.
